hpx/custom_exception_info.hpp

```diff
@@ -378,9 +100,52 @@ namespace hpx
         // Report an early or late exception and locally exit execution. There
         // isn't anything more we could do. The exception will be re-thrown
         // from hpx::init
-        HPX_EXPORT void report_exception_and_continue(std::exception_ptr const&);
+        HPX_EXPORT void report_exception_and_continue(
+            std::exception_ptr const&);
         HPX_EXPORT void report_exception_and_continue(hpx::exception const&);
-    }
+
+        template <typename Exception>
+        HPX_EXPORT std::exception_ptr construct_exception(Exception const& e,
+            std::string const& func, std::string const& file, long line,
+            std::string const& back_trace, std::uint32_t node,
+            std::string const& hostname, std::int64_t pid, std::size_t shepherd,
+            std::size_t thread_id, std::string const& thread_name,
+            std::string const& env, std::string const& config,
+            std::string const& state_name, std::string const& auxinfo)
+        {
+            // create a std::exception_ptr object encapsulating the Exception to
+            // be thrown and annotate it with all the local information we have
+            try
+            {
+                throw_with_info(e,
+                    std::move(hpx::exception_info().set(
+                        hpx::detail::throw_stacktrace(back_trace),
+                        hpx::detail::throw_locality(node),
+                        hpx::detail::throw_hostname(hostname),
+                        hpx::detail::throw_pid(pid),
+                        hpx::detail::throw_shepherd(shepherd),
+                        hpx::detail::throw_thread_id(thread_id),
+                        hpx::detail::throw_thread_name(thread_name),
+                        hpx::detail::throw_function(func),
+                        hpx::detail::throw_file(file),
+                        hpx::detail::throw_line(line),
+                        hpx::detail::throw_env(env),
+                        hpx::detail::throw_config(config),
+                        hpx::detail::throw_state(state_name),
+                        hpx::detail::throw_auxinfo(auxinfo))));
+            }
+            catch (...)
+            {
+                return std::current_exception();
+            }
+
+            // need this return to silence a warning with icc
+            HPX_ASSERT(false);
+            return std::exception_ptr();
+        }
```

Comment:
Is it necessary to _define_ this in a header file?

Reply:
Well, It *is* a templated function, but we might be able to get by with explicit instantiations. Not sure.

Reply:
We do explicitly instantiate this function (see: https://github.com/STEllAR-GROUP/hpx/blob/master/src/exception.cpp#L412-L461).

Reply:
Sorry. Then this can definitely just be declared. Thanks!

Reply:
Looking a bit more carefully I see that it's `throw_exception` that's explicitly instantiated, not `construct_exception`. We would probably manage with the same explicit instantiations for `construct_exception` as well. Would you prefer that?

Reply:
Yes, you are right. However, `construct_exception` is instantiated by `get_exception`, which is implicitly instantiated by `throw_exception` so there is no need to explicitly instantiate that as well.

Reply:
Sorry for the confusion (I'm slowly remembering the changes I made). That part has changed on this branch. `throw_exception` does not call `construct_exception` anymore. We can still instantiate `construct_exception` explicitly though.

Reply:
Right, I should have looked more carefully. Let's go with explicitly instantiating the functions, then.